src/main/java/frc/robot/poseestimator/helpers/ObservationCountHelper.java

```diff
@@ -0,0 +1,38 @@
+package frc.robot.poseestimator.helpers;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+public class ObservationCountHelper<T> {
+
+	private final Supplier<List<T>> observationSupplier;
+	private final int maxDataIntakeRepeats;
+	private List<T> stackedObservations;
+	private int dataIntakeCounter;
+
+	public ObservationCountHelper(Supplier<List<T>> observationSupplier, int maxCount) {
+		this.observationSupplier = observationSupplier;
+		this.maxDataIntakeRepeats = maxCount;
+		stackedObservations = new ArrayList<>();
+		dataIntakeCounter = 0;
+	}
+
+	public List<T> getStackedObservations() {
+		if (dataIntakeCounter == maxDataIntakeRepeats) {
+			dataIntakeCounter = 0;
+			List<T> copyOfStackedObservations = stackedObservations;
+			stackedObservations = new ArrayList<>();
+			return copyOfStackedObservations;
+		} else {
+			dataIntakeCounter++;
+			stackedObservations.addAll(observationSupplier.get());
+			return new ArrayList<>();
+		}
+	}
```

Comment:
why? you wrote it like a do-while.
when the last observation is added, you need to wait to the next cycle 
to only get that List when you had the entire list the previous cycle.

make it so it first adds the observation and then checks if it reached max count